### PR TITLE
S3 object store

### PR DESCRIPTION
**Pull request by ravi-mosaicml**
This PR is the stacked diff for https://github.com/mosaicml/composer/pull/1147

### CODE REVIEW

File: composer/utils/object_store/s3_object_store.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+    def download_object(
+        self,
+        object_name: str,
+        filename: Union[str, pathlib.Path],
+        overwrite: bool = False,
+        callback: Optional[Callable[[int, int], None]] = None,
+    ):
+        if os.path.exists(filename):
+            raise FileExistsError(f"The file at {filename} already exists")
```
Comment (abhi-mosaic) [inline composer/utils/object_store/s3_object_store.py]: Curious, should this still occur if `overwrite=True`?

File: setup.py [hunk 2]
```diff
@@ -121,6 +121,7 @@ def package_files(prefix: str, directory: str, extension: str):
     "pandoc==2.2",
     "pypandoc==1.8.1",
     "GitPython==3.1.27",
+    "moto[s3]>=3.1.12,<3.2",
```
Comment (abhi-mosaic) [inline setup.py:124]: boto :)

**Review by abhi-mosaic — APPROVED** [review]
LGTM, left a comment re. `overwrite=True`

File: composer/utils/object_store/s3_object_store.py (hunk 1 above)
Comment (ravi-mosaicml) [reply composer/utils/object_store/s3_object_store.py]: Whoopsies, thanks.

Comment (ravi-mosaicml) [reply composer/utils/object_store/s3_object_store.py]: Fixed

**Pull request closed without merging**